### PR TITLE
Specify restricted address range for precompiles/system contracts

### PR DESCRIPTION
**Pull request by axic**
Specify an Ethereum address range occupied by precompiles and future system contracts. Regular accounts and contracts cannot obtain such an address.

### CODE REVIEW

File: EIPS/eip-draft_restrict_precompile_address_range.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+---
+
+## Simple Summary
+Specify an Ethereum address range occupied by precompiles and future system contracts. Regular accounts and contracts cannot obtain such an address.
+
+## Abstract
+The address range between 0x0000000000000000000000000000000000000000 and 0x00000000000000000000000000000000000000ff is reserved for precompiles and system contracts.
+
+## Motivation
+This will simplify certain future features where unless this is implemented, several exceptions must be specified.
```
Comment (Arachnid) [inline EIPS/eip-draft_restrict_precompile_address_range.md]: Can you expand on this? What will it simplify?

File: EIPS/eip-draft_restrict_precompile_address_range.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+## Specification
+The address range between 0x0000000000000000000000000000000000000000 and 0x00000000000000000000000000000000000000ff is reserved for precompiles and system contracts.
+
+If a contract creation (as a result of a create transaction or a `CREATE` opcode) results in an address within this range, then it is rejected.
```
Comment (Arachnid) [inline EIPS/eip-draft_restrict_precompile_address_range.md]: Is this even worth checking for?

File: EIPS/eip-draft_restrict_precompile_address_range.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+## Rationale
+N/A
+
+## Backwards Compatibility
+No contracts on the main network have been created at the specified addresses. As a result it should pause no backwards compatibility problems.
```
Comment (Arachnid) [inline EIPS/eip-draft_restrict_precompile_address_range.md]: s/pause/pose/

File: EIPS/eip-draft_restrict_precompile_address_range.md (hunk 2 above)
Comment (axic) [reply EIPS/eip-draft_restrict_precompile_address_range.md]: If it is not specified then clients will introduce random assumptions. Why is it better not to specify it even if probability of extremely low?

File: EIPS/eip-draft_restrict_precompile_address_range.md (hunk 1 above)
Comment (axic) [reply EIPS/eip-draft_restrict_precompile_address_range.md]: Examples:
- reducing call cost for precompiles
- "future system contracts" can be specified much simpler if they can refer to "precompiles" without the need to specify all their addresses (which is a moving target)

**Comment by axic** [comment]
@Arachnid while I am happy to discuss changes here I am a bit confused about the EIP process. I though the idea now is that these EIP proposals are merged quickly and then they are discussed under the URL pointed to be `discussions-to`.

File: EIPS/eip-draft_restrict_precompile_address_range.md (hunk 1 above)
Comment (axic) [reply EIPS/eip-draft_restrict_precompile_address_range.md]: I actually had the idea of splitting it into two:
- 0x0 to 0xff for precompiles
- 0x100 to 0x1ff for "system contracts", which have code in the state (e.g. blockhash refactoring or (the now obsolete) casper contract).

Comment (axic) [reply EIPS/eip-draft_restrict_precompile_address_range.md]: I would like to make a distinction between precompiles (no code in state) and system contracts (code in state). What ranges to you propose for these two?

File: EIPS/eip-draft_restrict_precompile_address_range.md (hunk 2 above)
Comment (axic) [reply EIPS/eip-draft_restrict_precompile_address_range.md]: One reason against adding that clause is that there doesn't seem to be an easy way to add a test for it without changing internals :)

File: EIPS/eip-draft_restrict_precompile_address_range.md (hunk 1 above)
Comment (jakelang) [reply EIPS/eip-draft_restrict_precompile_address_range.md]: @axic considering that the definition of a "precompile" may be revised a bit if we introduce wasm precompiles that will likely be more numerous, then i would say that the reserved address range should be at least 65536 addresses long.

This also depends on whether we want to reserve addresses for wasm precompiles, or just consider them a different class of contract in the state.

Comment (axic) [reply EIPS/eip-draft_restrict_precompile_address_range.md]: Changed the reserved range.

File: EIPS/eip-draft_restrict_precompile_address_range.md (hunk 2 above)
Comment (axic) [reply EIPS/eip-draft_restrict_precompile_address_range.md]: Please clarify: is this a comment blocking the merging of this draft or should this properly be evaluated on the discussion url?

I could also see a middle ground by stating: due to the low probability (and lack of adequate testing possibilities) no explicit checks should be added to ensure that external transaction signing or the invoking of the `CREATE` instruction can result in a precompile address.

**Comment by axic** [comment]
Updated to reflect the comments. What is left to merge this?

**Comment by axic** [comment]
@Arachnid @Souptacular what is left to merge this as draft?

**Comment by Souptacular** [comment]
@axic You are correct that well formed EIPs are to be merged as draft and then discussed at the discussion URL. Merging.

**Comment by axic** [comment]
Since this draft was merged, please take all the discussion to https://ethereum-magicians.org/t/eip-1352-specify-restricted-address-range-for-precompiles-system-contracts/1151.

**Pull request merged**